### PR TITLE
Fix TP sharding issue

### PR DESCRIPTION
**Pull request by RissyRan**
# Description

Fix sharding shape misalignment issue for TP & TP transpose.

# Tests

Before the change:

1) flash attention meets ` assert sinks.shape == (num_q_heads,)` issue
2) MLP met `intermediate_output = intermediate_output + wo_bias` misalignement issue
3) Shape imcompatible issue during decoding due to padding (variable overwritten) 

After the change:
* flash attention: [link](https://paste.googleplex.com/5469053783244800)
* MLP: [link](https://paste.googleplex.com/6526021544968192)
* Decoding without any shape compatible issue: [link](https://paste.googleplex.com/6278671895363584)

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

**Review by shuningjin — APPROVED** [review]
Thank you for the fix!

File: src/MaxText/layers/moe.py [hunk 1]
```diff
@@ -797,14 +798,14 @@ def gmm(inputs, kernel, group_sizes, expert_assignments):
         if isinstance(kernel, aqt.QTensor):
           # Multiply outputs by the kernely scale
           scales = jnp.take(kernel.scale[0].squeeze(), indices=expert_assignments, axis=0)
-          if hs_shape[0] % pad_length:
+          if padding_amount > 0:
             scales = jax.lax.pad(
                 scales,
                 jnp.array(0.0, dtype=scales.dtype),
                 [(0, pad_length, 0), (0, 0, 0)],
```
Comment (gagika) [inline src/MaxText/layers/moe.py]: Shouldn't it be padding_amount instead of pad_length inside if?

```
if padding_amount > 0:
    scales = jax.lax.pad(
        scales, jnp.array(0.0, dtype=scales.dtype),
        [(0, padding_amount, 0), (0, 0, 0)],
    )
 
 ```

File: src/MaxText/layers/moe.py [hunk 2]
```diff
@@ -747,9 +747,10 @@ def gmm(inputs, kernel, group_sizes, expert_assignments):
...
         raise ValueError("The number of input tokens must match the number of expert" " assignments!")
+      padding_amount = 0
       if hs_shape[0] % pad_length:
-        pad_length = pad_length - hs_shape[0] % pad_length
-        inputs = jax.lax.pad(inputs.astype(jnp.float32), 0.0, [(0, pad_length, 0), (0, 0, 0)])
+        padding_amount = pad_length - hs_shape[0] % pad_length
+        inputs = jax.lax.pad(inputs.astype(jnp.float32), 0.0, [(0, padding_amount, 0), (0, 0, 0)])
 
       inputs = inputs.astype(self.dtype)
       kernel = kernel.astype(self.dtype)
```
Comment (gagika) [inline src/MaxText/layers/moe.py]: is there a reason for padding to happen in float32 then casing to self.dtype?

if not, we can do in `self.dtype` with something like this:

```
inputs = inputs.astype(self.dtype)
kernel = kernel.astype(self.dtype)

if hs_shape[0] % pad_length:
    padding_amount = pad_length - hs_shape[0] % pad_length
    zero = jnp.array(0, dtype=inputs.dtype)
    inputs = jax.lax.pad(inputs, zero, [(0, padding_amount, 0), (0, 0, 0)])
```

File: src/MaxText/layers/moe.py (hunk 1 above)
Comment (RissyRan) [reply src/MaxText/layers/moe.py]: Good catch for the quantized part. Updated.

File: src/MaxText/layers/moe.py (hunk 2 above)
Comment (RissyRan) [reply src/MaxText/layers/moe.py]: Good catch! Updated to `inputs = jax.lax.pad(inputs, jnp.array(0.0, dtype=inputs.dtype), [(0, padding_amount, 0), (0, 0, 0)])` and tested end-to-end as well.
